packages/lisk-p2p/src/peer.ts

```diff
@@ -44,34 +58,144 @@ export interface PeerConnectionState {
+				this.emit(EVENT_INVALID_REQUEST_RECEIVED, request);
+
+				return;
+			}
+			if (
+				request.procedure === 'updateMyself' &&
+				typeof request.data === 'object'
+			) {
+				// Internal handling of request to extract the PeerInfo.
+				this._handlePeerInfo(request);
+			}
+			// Emit request for external use.
+			this.emit(EVENT_REQUEST_RECEIVED, request);
+		};
+
+		// This needs to be an arrow function so that it can be used as a listener.
+		this._handleMessage = (packet: unknown) => {
+			// TODO later: Switch to LIP protocol format.
+			// TODO ASAP: Move validation/sanitization to a separate file with other validation logic.
+			const message = packet as ProtocolInboundMessage;
+			if (!message || typeof message.event !== 'string') {
+				this.emit(EVENT_INVALID_MESSAGE_RECEIVED, message);
+
+				return;
+			}
+			this.emit(EVENT_MESSAGE_RECEIVED, message);
+		};
 	}
 
-	private _createOutboundSocket(): any {
-		if (!this._outboundSocket) {
-			this._outboundSocket = socketClusterClient.create({
-				hostname: this._ipAddress,
-				port: this._wsPort,
-				query: this._nodeStatus,
-				autoConnect: false,
-			});
+	public get height(): number {
+		return this._height;
+	}
+
+	public get id(): string {
+		return this._id;
+	}
+
+	public set inboundSocket(value: any) {
+		if (this._inboundSocket) {
+			this._stopListeningForRPCs(this._inboundSocket);
+			this._stopListeningForMessages(this._inboundSocket);
 		}
+		this._inboundSocket = value;
+		this._startListeningForRPCs(this._inboundSocket);
+		this._startListeningForMessages(this._inboundSocket);
+	}
+
+	public get inboundSocket(): any {
+		return this._inboundSocket;
+	}
+
+	public get ipAddress(): string {
+		return this._ipAddress;
+	}
+
+	public get nodeInfo(): P2PNodeInfo | undefined {
+		return this._nodeInfo;
+	}
+
+	public set outboundSocket(value: any) {
+		this._outboundSocket = value;
+	}
+
+	public get outboundSocket(): any {
+		return this._outboundSocket;
+	}
+
+	public get peerInfo(): PeerInfo {
+		return this._peerInfo;
+	}
+
+	public get state(): PeerConnectionState {
+		const inbound = this._inboundSocket
+			? this._inboundSocket.state === this._inboundSocket.OPEN
+				? ConnectionState.CONNECTED
+				: ConnectionState.DISCONNECTED
+			: ConnectionState.DISCONNECTED;
+		const outbound = this._outboundSocket
+			? this._outboundSocket.state === this._outboundSocket.OPEN
+				? ConnectionState.CONNECTED
+				: ConnectionState.DISCONNECTED
+			: ConnectionState.DISCONNECTED;
+
+		return {
+			inbound,
+			outbound,
+		};
+	}
+
+	public get wsPort(): number {
+		return this._wsPort;
+	}
+
+	public static constructPeerId(ipAddress: string, port: number): string {
+		return `${ipAddress}:${port}`;
+	}
+
+	/**
+	 * This is not a declared as a setter because this method will need
+	 * invoke an async RPC on the socket to pass it the new node status.
+	 */
+	public applyNodeInfo(nodeInfo: P2PNodeInfo): void {
+		this._nodeInfo = nodeInfo;
```

Comment:
do we need to maintain `nodeInfo` per Peer?

Reply:
Yes because we need to pass it as a query parameter as part of the WebSocket connection (required by current protocol).